### ISSUE TITLE
Verify behaviour when license expires

### ISSUE THREAD

**Issue by knolleary** [flowforge/flowforge#1739]
As part of the improving licensing handling (#1638) we need to verify our existing behaviour when a license expires to ensure it aligns with what we want to happen in terms of business requirements.

We currently only check the validity of the license on startup. So a long running system wouldn't notice the license expiring.

At a minimum, we need to log the fact the license has expired when it happens.

We then need to decide on the desired behaviour.
 - Does everything keep running?
 - Do we block creation of any new projects?
 - Do EE-only features continue working?

Given all of our logic around EE-only features is driven on startup, I don't think we have the ability to turn EE features off without a restart.

The telemetry reporting should also indicate if the license has expired somehow.

**Comment by hardillb** [comment]
Big banner on every page for a period of time (2 weeks?) then stop creating new projects, existing projects continue to run (possibly no suspend/restart after suspend until back under default limits)

**Comment by ZJvandeWeg** [comment]
We should reach out 90 days before to renew, that being said: maybe we provide them with 30 days of warning after and block all access after, and suspend all projects? 

And display a screen to capture the new license?

**Comment by knolleary** [comment]
Based on comments above, this is a proposed set of behaviours:

1. We, FF Inc., reach out to customers with 90 days notice.
2. 30 days prior to expiry
    1. Platform sends email to all Admin users notifying them of pending license expiry
    2. Add banner for Admin users to notify the license is nearing expiry. I don't think we should add banner for non-admin users as they are not able to do anything about it.
3. Day of expiry
    1. platform sends email to all Admin users notifying them of expiry 
4. Have 30 days grace period after license expiry. At this point, adding a banner for *all* users is probably a suitable escalation.
5. At end of grace period:
    1. Platform sends email to all Admins to notify them
    2. Suspend all projects
    3. Do *something* with devices. The only option we have today is to unassign all devices from projects so they stop running... but would then be up to the users to reassign once a license is applied. That isn't ideal - as it makes recovery harder for the user.
    4. Expire all non-admin login sessions
    5. Admin is only able to access the licensing page to provide a new license.

---

What this doesn't allow for is a licensed user deciding to fall back to the pure OSS platform they can run without a license. It suggests once licensed, always licensed.

Whilst we don't necessarily need to have a fully automated solution for that, it feels like we should have some path, even if it involves some amount of manual work to migrate across.

**Comment by hardillb** [comment]
We could solve the fall back to CE/OSS by having a delete license button in the admin interface to remove the expired license from the DB, it would then act as CE/OSS on restart.

Would still need to work out what to do with items over CE/OSS limit (but refusing to create more until back under limit feels right)

**Comment by Steve-Mcl** [comment]
> 2. 30 days prior to expiry
>     i. Platform sends email to all Admin users notifying them of pending license expiry
>     ii. Add banner for Admin users to notify the license is nearing expiry. I don't think we should add banner for non-admin users as they are not able to do anything about it.

Regarding the "Add banner for Admin users" part, does this include Team Owners? I feel an admin may not login very often however a team owner is more likely to login. 

Maybe consider a "non-dismissible banner" for Admins and a "dismissible banner" for ALL non-admins. This could be shown at every login and / or a full page refresh?

<br>

> Expire all non-admin login sessions (and prevent non-admin users from logging in)

Should the non-admins failed login message indicate why? Perhaps this will prompt them to "get this sorted out"?

<br>

> Do something with devices. The only option we have today is to unassign all devices from projects so they stop running... but would then be up to the users to reassign once a license is applied. That isn't ideal - as it makes recovery harder for the user.

Yuk.  I would prefer not to do this - perhaps there is an simple solution in the "Agent Manager" we could explore?

<br>

> We could solve the fall back to CE/OSS by having a delete license button in the admin interface to remove the expired license from the DB, it would then act as CE/OSS on restart.

Seems sane & easy enough to achieve (could also force a restart after removal?)

<br>


> Would still need to work out what to do with items over CE/OSS limit (but refusing to create more until back under limit feels right)

The Audit logs will (after PR merged) on a weekly basis remind the admin what & where they are over limits. Agree with refusal to create more until levels are below limits.

**Comment by ZJvandeWeg** [comment]
Leaning on your experience @Steve-Mcl, not sure why one could unsuspend everything for 2 hours, but I guess it's senseable not to stop the line. I'm OK with your updated proposal.

**Comment by knolleary** [comment]
> Allow any/all projects to be un-suspended for a period (2h is common) with full EE experience after the 2h, projects will be again suspended.

I think that is making it unnecessarily complicated - the work required to implement that cleanly will be substantial and isn't really warranted for this edge case.

My proposal was based on the approach of 'given them 90 days notice, 30 days grace, then shutdown' - so a total of 4 months warning, which does feel fair.

I'm assuming your suggested alternative is based on not losing access to production systems in the worst case scenario? If that is a concern, then the most simple alternative would be for the 30 days grace to end without removing any functionality or removing access, but adding banners *everywhere* to make it clear the platform is unlicensed.

Given telemetry is becoming a required feature of licensed platforms, we'd know full well they were running without license for an extended period. If they did anything to block the telemetry call home (but had it appear to work), then they won't get any support from us.

**Comment by Steve-Mcl** [comment]
From comments in PR...

>> consider making a user config option?

> No

>> Should the license itself specify this? (billable grace period?)

> No

>> license_expiry_stops_all is set to false.
>> Do we want to enable this?
>> Should the license itself specify this? (billable feature?)

> This is the discussion in https://github.com/flowforge/flowforge/issues/1739 which is where any other proposals should be discussed. I don't want one set of things agreed in the issue and then a different set of things implemented in the PR.

>> In the issue I think i wrote "yuk" to the proposal (sorry 😉) -

> The proposal was based on what we could do today. It made it clear there were problems with that approach.
> However, the proposal also changed to keep things running after the grace period. In which case we don't need to do anything with devices.

---


ok, cool - then these 3 points are already satisfied in PR & it is ready for review (once tests complete)

**Comment by knolleary** [comment]
Have reviewed the PR.

If the conclusion in this issue is that we keep everything running post-expiry, but add a banner and nagging email, then the summary of the full expected behaviour is:

1. We, FF Inc., reach out to customers with 90 days notice.
2. 30 days prior to expiry
    1. Platform sends email to all Admin users notifying them of pending license expiry
    2. **NEW** resend the email at 1 week intervals
    3. Add banner for Admin users to notify the license is nearing expiry. I don't think we should add banner for non-admin users as they are not able to do anything about it.
4. Day of expiry
    1. platform sends email to all Admin users notifying them of expiry 
    2. **NEW** Add a banner for *all* users that the license has expired
5. Every 5? days post expiry
    1. **NEW** platform sends email to all Admin users notifying them of expiry 

In summary - the post-expiry grace period doesn't serve much useful purpose - nothing different happens at the end of the grace period now.

**Comment by Steve-Mcl** [comment]
> 4. Every 5? days post expiry
>    1. **NEW** platform sends email to all Admin users notifying them of expiry

Perhaps weekly (same day) would make sense?

> In summary - the post-expiry grace period doesn't serve much useful purpose - nothing different happens at the end of the grace period now.

I thought we might use the grace period as "you have 30 days support grace"?  If not, I will ensure a setting of `0` means the license goes directly from warning to expired.

**Comment by knolleary** [comment]
> I thought we might use the grace period as "you have 30 days support grace"?

That's a business decision that doesn't have any bearing on the functionality of the code. If we don't have a grace period, we shouldn't have lots of code implementing it.

**Comment by knolleary** [comment]
The decision from the product review meeting was to keep this simple for now given we are still months away from anyone being in this situation and our requirements may well change in that time.

So the proposal is pretty much as described in my last comment - https://github.com/flowforge/flowforge/issues/1739#issuecomment-1447887543

**Comment by Steve-Mcl** [comment]
@knolleary 

All changes are in & a final review is now possible (tests still running as I write)

### Summary...

* Platform sends email to all Admin users notifying them of pending license expiry
   * resend the email at 1 week intervals
* Add banner for Admin users to notify the license is nearing expiry.
* Day of expiry:
  * platform sends email to all Admin users notifying them of expiry
  * Add a banner for all users that the license has expired
* Every Sunday when the license check reruns (2am - ready for monday morning email catchup)
  * platform sends email to all Admin users notifying them of expiry
